File: bench/binary_trees.ts
// TypeScript peer of bench/binary_trees.vader. Same recursive build +
// count algorithm. Node holds value + left/right refs, exactly mirroring
// the Vader struct.

const DEPTH = 17;

class Node {
  constructor(
    public readonly value: number,
    public readonly left: Node | null,
    public readonly right: Node | null,
  ) {}
}

function buildTree(depth: number, value: number): Node {
  if (depth === 0) return new Node(value, null, null);
  return new Node(value, buildTree(depth - 1, value * 2), buildTree(depth - 1, value * 2 + 1));
}

function countNodes(t: Node | null): number {
  if (t === null) return 0;
  return 1 + countNodes(t.left) + countNodes(t.right);
}

const tree = buildTree(DEPTH, 1);
const count = countNodes(tree);
console.log(`binary_trees depth=${DEPTH} count=${count}`);

File: bench/binary_trees.java
// Java peer of bench/binary_trees.vader. `Node` is a record (Java 14+),
// holding value + left/right refs ; null is the empty subtree, matching
// Vader's `Node | null`.

public class binary_trees {
    static final int DEPTH = 17;

    record Node(int value, Node left, Node right) {}

    static Node buildTree(int depth, int value) {
        if (depth == 0) return new Node(value, null, null);
        return new Node(value,
            buildTree(depth - 1, value * 2),
            buildTree(depth - 1, value * 2 + 1));
    }

    static int countNodes(Node t) {
        if (t == null) return 0;
        return 1 + countNodes(t.left) + countNodes(t.right);
    }

    public static void main(String[] args) {
        Node tree = buildTree(DEPTH, 1);
        int count = countNodes(tree);
        System.out.printf("binary_trees depth=%d count=%d%n", DEPTH, count);
    }
}
